### PR TITLE
feat: implement config package

### PR DESCRIPTION
**Pull request by PierreDemailly**
TODOs

- [x] Generics
- [x] More tests
- [x] Do we need async methods ? The package is full sync

I would appreciate a first review @fraxken

### CODE REVIEW

**Comment by fraxken** [comment]
I would have re-implemented it with async and the old test suite (not perfect but it was doing the job). Synchronous API are not idiot but require a deeper refactor.

And yes Generics is welcome

**Comment by PierreDemailly** [comment]
I've added some test and the coverage is now 97%

@fraxken  WDYM by deeper refactor ? I'd like to implement both sync and async methods (`read`, `readSync`, `writeOnDisk`, `writeOnDiskSync` etc), do you think its better to skip sync and keep only async ?

**Comment by fraxken** [comment]
@PierreDemailly Some API has been designed to be lazy with async. Also having both make things a lot more complicated for me in term of how things are abstract.

File: src/config/README.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
+<p align="center"><h1 align="center">
+  Config
+</h1>
+
+<p align="center">
+  Reactive configuration loader with safe hot reload configuration upon [JSON Schema](https://json-schema.org/)
```
Comment (fraxken) [inline src/config/README.md:6]:
```suggestion
  Reactive configuration loader with safe hot reload configuration upon <a href="https://json-schema.org/" target="_blank">JSON Schema</a>
```

File: src/config/README.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+```
+
+Available options are:
+
+| name | type | default value | description |
+| --- | --- | --- | --- |
+| `createOnNoEntry` | `boolean` | `false` | Create the file with default payload value if he doesn't exist on the local disk |
+| `writeOnSet` | `boolean` | `false` | Write the file on the disk after each time `.set()` is called |
+| `autoReload` | `boolean` | `false` | Setup hot reload of the configuration file |
+| jsonSchema | `object` | `null` | The default JSON Schema for the configuration |
```
Comment (fraxken) [inline src/config/README.md:90]:
```suggestion
| `jsonSchema` | `object` | `null` | The default JSON Schema for the configuration |
```

File: src/config/README.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+> [!NOTE]
+> When no schema is provided, it will search for a file prefixed by `.schema` with the same config name.
+
+### `Config.read(defaultPayload?: object): void`
+
+Will read the local configuration on disk. A default `payload` value can be provided in case the file doesn't exist !
+
+> [!CAUTION] When the file doesn't exist, the configuration is written at the next loop iteration (with `lazyWriteOnDisk`).
```
Comment (fraxken) [inline src/config/README.md:99]:
```suggestion
> [!CAUTION]
> When the file doesn't exist, the configuration is written at the next loop iteration (with `lazyWriteOnDisk`).
```

File: src/config/src/utils.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  // eslint-disable-next-line no-param-reassign
+  const subDepth = --depth;
+  for (const [key, value] of Object.entries(obj)) {
+    Reflect.set(obj, key, limitObjectDepth(value, subDepth));
+  }
+
+  return obj as T;
+}
+
+export function objectValueByPath<T = any>(obj: Record<string, any>, path: string): T | null {
```
Comment (fraxken) [inline src/config/src/utils.ts:36]: J'aurais certainement adopté un nommage plus simple genre deepGet et deepSet ? Avec des exemples en JSDoc comme ça c'est clair pour quelqu'un qui vient lire le code.

File: src/config/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+  type: "object",
+  additionalProperties: true
+};
+
+export interface ConfigOptions<T> {
+  createOnNoEntry?: boolean;
+  autoReload?: boolean;
+  writeOnSet?: boolean;
+  jsonSchema?: JSONSchemaType<T>;
+}
```
Comment (fraxken) [inline src/config/src/index.ts:32]: I'll personally add an option to inject FS to simplify testing

**Comment by fraxken** [comment]
I would rename Config -> ConfigSync and save it into a separate files. And keep index to export.

File: src/config/src/index.ts (hunk 5 above)
Comment (PierreDemailly) [reply src/config/src/index.ts:32]: Mmh, the only test (i think) that would benefit ioc is this one

```ts
it("should recreate file on read when file does not exists and createOnNoEntry is true", async() => {
  const configPath = path.join(__dirname, "fixtures", ".doesNotExists3");
  assert(fs.existsSync(configPath) === false);

  const config = new Config(configPath, { createOnNoEntry: true });
  await config.read({ boo: "boom" });
  await once(config, "configWritten");

  assert(fs.existsSync(configPath) === true);
  assert.deepEqual(config.payload, { boo: "boom" });

  fs.rmSync(configPath);
  assert(fs.existsSync(configPath) === false);

  await config.read();
  await once(config, "configWritten");
  assert(fs.existsSync(configPath) === true);
  assert.deepEqual(config.payload, { boo: "boom" });

  await config.close();
  fs.rmSync(configPath);
});
```

Dealing with a fs mock would essentially be the same thing/complexity, wouldn't it?

Comment (fraxken) [reply src/config/src/index.ts]: Could be useful outside of the project (like in RC or other projects)

**Review by fraxken — APPROVED** [review]

**Comment by fraxken** [comment]
@allcontributors please add @PierreDemailly for code, docs, maintenance

**Pull request merged**